File: localink_be/Services/Implementations/LocationService.cs
using Microsoft.Extensions.Caching.Memory;
using System.Net.Http;

public class LocationService : ILocationService
{
    private readonly HttpClient _http;
    private readonly IMemoryCache _cache;
    private readonly string apiKey;

    // CACHE
    private static Dictionary<string, string> cache = new();

    public LocationService(HttpClient http, IConfiguration config, IMemoryCache cache)
    {
        _http = http;
        _cache = cache;

        apiKey = config["CSC:ApiKey"] ?? throw new Exception("API Key missing");
        _http.DefaultRequestHeaders.Add("X-CSCAPI-KEY", apiKey);
    }

    private async Task<string> GetCached(string url)
    {
        if (_cache.TryGetValue(url, out string? cached) && cached != null)
            return cached;

        try{
            var res = await _http.GetStringAsync(url);
            _cache.Set(url, res, TimeSpan.FromHours(24)); // cache 24 hrs
            return res;
        }
        catch (Exception ex)
        {
            throw new Exception("Location API failed", ex);
        }
    }

    public async Task<string> GetCountries()
    {
        return await GetCached("https://api.countrystatecity.in/v1/countries");
    }

    public async Task<string> GetStates(string countryCode)
    {
        return await GetCached($"https://api.countrystatecity.in/v1/countries/{countryCode}/states");
    }

    public async Task<string> GetCities(string countryCode, string stateCode)
    {
        return await GetCached($"https://api.countrystatecity.in/v1/countries/{countryCode}/states/{stateCode}/cities");
    }
}

File: localink-fe/src/app/services/location.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class LocationService {

  private baseUrl = '/api/location'; // proxy will handle backend URL

  constructor(private http: HttpClient) {}

  getCountries(): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/countries`);
  }

  getStates(countryCode: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/states/${countryCode}`);
  }

  getCities(countryCode: string, stateCode: string): Observable<any[]> {
    return this.http.get<any[]>(
      `${this.baseUrl}/cities/${countryCode}/${stateCode}`
    );
  }
}